File: ios/Counter.m
//
//  Counter.m
//  ReactNativeDemo
//
//  Created by Jacob Losin on 4/24/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

// First argument is the name of the Swift class, the second is it's superclass
@interface RCT_EXTERN_MODULE(Counter, RCTEventEmitter)

RCT_EXTERN_METHOD(increment)
RCT_EXTERN_METHOD(getCount: (RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(getMultipleArguments: (RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(
                  decrement: (RCTPromiseResolveBlock)resolve
                  rejecter: (RCTPromiseRejectBlock)reject
                  )


@end

/*
 If you want to expose the module with a different name other than 'Counter'...
 
 @interface RCT_EXTERN_REMAP_MODULE(RNCounter, Counter, NSObject)
 @end
 
 The module will now be exposed as 'RNCounter'
 */


/*
 If you want to rename a function...
 // 'false' means the method is not synchronous
 @interface _RCT_EXTERN_REMAP_METHOD(inc, increment, false)
 @end
 */

File: ios/Counter.swift
//
//  Counter.swift
//  RNDemo
//
//  Created by Jacob Losin on 5/23/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

@objc(Counter)
class Counter: RCTEventEmitter {
  
  private var count = 0
  
  // Suppresses the 'requiresMainQueueSetup' warning
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  override func constantsToExport() -> [AnyHashable : Any]! {
    return [
      "number": 123.9,
      "string": "foo",
      "boolean": true,
      "array": [1, 22.2, "33"],
      "object": ["a": 1, "b": 2]
    ]
  }
  
  @objc
  func increment() {
    count += 1
    //print("count is \(count)")
    
    sendEvent(withName: "onIncrement", body: ["count": count])
  }
  
  override func supportedEvents() -> [String]! {
    return ["onIncrement"]
  }
  
  //Return a single value
  @objc
  func getCount(_ callback:RCTResponseSenderBlock) {
    callback([count])
  }
  
  //Return multiple values
  @objc
  func getMultipleArguments(_ callback: RCTResponseSenderBlock) {
    callback([
      count,               // variable
      123.9,               // int or float
      "third argument",    // string
      [1, 2.2, "3"],       // array
      ["a": 1, "b": 2]     // object
      ])
  }
  
  //Promises
  @objc
  func decrement(
    _ resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
    if (count == 0) {
      let error = NSError(domain: "", code: 200, userInfo: nil)
      reject("E_COUNT", "count cannot be negative", error)
    } else {
      count -= 1
      resolve("count was decremented")
    }
  }
  
}
